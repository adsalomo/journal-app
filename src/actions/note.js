import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from '../types/types';

export const startNewNote = () => {
    return async (dispatch, getState) => {

        const uid = getState().auth.uid;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime(),
        };

        const doc = await db.collection(`${uid}/journal/note`).add(newNote);

        dispatch(activeNote(doc.id, newNote));
        dispatch(addNewNote(doc.id, newNote));
    };
};

export const activeNote = (id, note) => {
    return {
        type: types.noteActive,
        payload: {
            id,
            ...note,
        }
    };
};

export const addNewNote = (id, note) => {
    return {
        type: types.noteAddNew,
        payload: {
            id, ...note,
        },
    }
};

export const startLoadingNote = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    };
}

export const setNotes = (notes) => {
    return {
        type: types.noteLoad,
        payload: notes,
    };
};

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {

        const { uid } = getState().auth;

        // FIREBASE NO SOPORTA UNDEFINED, LA URL ESTA HASTA LA CLASE 263 SIN LA URL
        if (!note.url) {
            delete note.url;
        }

        const noteToFireStore = { ...note };
        delete noteToFireStore.id;

        await db.doc(`${uid}/journal/note/${note.id}`).update(noteToFireStore);

        dispatch(refreshNote(note.id, noteToFireStore));

        Swal.fire('Saved', note.title, 'success');
    }
};

export const refreshNote = (id, note) => {
    return {
        type: types.noteUpdated,
        payload: {
            id,
            note: {
                id,
                ...note
            }
        }
    }
};

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active: activeNote } = getState().note;

        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
        });

        Swal.showLoading();

        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;

        dispatch(startSaveNote(activeNote));

        Swal.close();
    }
};

export const startDeleting = (id) => {
    return async (dispatch, getState) => {

        const uid = getState().auth.uid;
        await db.doc(`${uid}/journal/note/${id}`).delete();

        dispatch(deleteNote(id));
    }
};

export const deleteNote = (id) => {
    return {
        type: types.noteDeleteNote,
        payload: id,
    }
};

export const noteLogout = () => {
    return {
        type: types.noteLogoutCleaning,
    }
};

