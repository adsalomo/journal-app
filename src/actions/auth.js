import { types } from '../types/types';
import { firebase, googleAuthProvide } from '../firebase/firebase-config';
import { uiFinishLoading, uiStartLoading } from './ui';
import Swal from 'sweetalert2'
import { noteLogout } from './note';

// middleware
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        dispatch(uiStartLoading());

        firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(uiFinishLoading());

                dispatch(login(user.uid, user.displayName));
            }, e => {
                dispatch(uiFinishLoading());
                console.log(e);
                Swal.fire('Error', e.message, 'error');
            });

    }
};

// middleware
export const startRegister = (name, email, password) => {
    return (dispatch) => {

        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({ displayName: name });
                dispatch(login(user.uid, user.displayName));
            }, e => {
                console.log(e);
                Swal.fire('Error', e.message, 'error');
            });

    }
};

// middleware
export const startGoogleLogin = () => {
    return (dispatch) => {

        firebase.auth().signInWithPopup(googleAuthProvide)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
            }, e => {
                console.log(e);
            });

    }
};

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName,
        }
    }
};

// middleware
export const startLogout = () => {
    return async (dispatch) => {

        await firebase.auth().signOut();
        dispatch(logout());
        dispatch(noteLogout());
    }
}

export const logout = () => {
    return {
        type: types.logout,
    }
}

