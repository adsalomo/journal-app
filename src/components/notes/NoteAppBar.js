import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploading } from '../../actions/note';

export const NoteAppBar = () => {

    const dispatch = useDispatch();
    const { active: note } = useSelector(state => state.note);

    const handleSave = () => {
        dispatch(startSaveNote(note));
    }

    const handleUploadFile = () => {
        document.querySelector("#fileSelector").click();
    }

    const handleOnChange = (e) => {
        const file = e.target.files[0];
        if (file) {
            dispatch(startUploading(file));
        }
    }

    return (
        <div className="note__app-bar">
            <span>28 agosto 2020</span>

            <input
                name="file"
                id="fileSelector"
                type="file"
                style={{ display: 'none' }}
                onChange={handleOnChange}
            />

            <div>
                <button
                    className="btn"
                    onClick={handleUploadFile}
                >
                    Picture
                </button>

                <button
                    className="btn"
                    onClick={handleSave}
                >
                    Save
                </button>
            </div>
        </div>
    )
}
