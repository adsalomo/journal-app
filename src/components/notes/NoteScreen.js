import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/note';
import { useForm } from '../../hooks/useForm';
import { NoteAppBar } from './NoteAppBar';

export const NoteScreen = () => {

    const { active: note } = useSelector(state => state.note);
    const dispatch = useDispatch();

    const [formValues, handleInputChange, reset] = useForm(note);
    const { title, body, id } = formValues;

    const activeId = useRef(note.id);

    useEffect(() => {

        if (note.id !== activeId.current) {
            reset(note);
            activeId.current = note.id;
        }

    }, [note, reset]);

    useEffect(() => {

        dispatch(activeNote(note.id, { ...formValues }));

    }, [formValues, dispatch]);

    const handleDelete = () => {
        dispatch(startDeleting(id))
    }

    return (
        <div className="note__main-content">

            <NoteAppBar />

            <div className="note__content">

                <input
                    className="note__title-input"
                    type="text"
                    name="title"
                    placeholder="Input"
                    autoComplete="off"
                    value={title}
                    onChange={handleInputChange}
                />

                <textarea
                    className="note__textarea"
                    placeholder="Escribe aquÃ­"
                    name="body"
                    value={body}
                    onChange={handleInputChange}
                ></textarea>

                {
                    note.url &&
                    <div className="note__image">
                        <img
                            src={note.url}
                            alt="imagen"
                        />
                    </div>
                }

            </div>

            <div
                className="btn btn-danger"
                onClick={handleDelete}
            >
                Delete
            </div>
        </div>
    )
}
