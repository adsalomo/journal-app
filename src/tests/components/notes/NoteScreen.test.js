import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { activeNote } from "../../../actions/note";
import { NoteScreen } from "../../../components/notes/NoteScreen";

jest.mock('../../../actions/note', () => {
    return {
        activeNote: jest.fn(),
    }
});

const middlewares = [thunk];
const mockStore = configureStore(middlewares);


const initialState = {
    auth: {
        uid: 'ABC',
        name: 'Elena'
    },
    note: {
        active: {
            id: 123,
            title: 'Hola mundo',
            body: 'Soy el cuerpo',
            date: 0,
        }
    }
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

describe('Pruebas NoteScreen', () => {

    const wrapper = mount(
        <Provider store={store}>
            <NoteScreen />
        </Provider>
    );

    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de disparar el activeNote', () => {
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'Nuevo titulo'
            }
        });

        expect(activeNote).toHaveBeenCalled();

        // Cuando el componente se crea, se ejecuta la funcion,
        // Luego cuando el se hace la simulacion, se ejecuta de nuevo la funcion
        // por eso debe de evaluarse con
        expect(activeNote).toHaveBeenLastCalledWith(
            123,
            {
                id: 123,
                title: 'Nuevo titulo',
                body: 'Soy el cuerpo',
                date: 0,
            }
        );

    });


});

