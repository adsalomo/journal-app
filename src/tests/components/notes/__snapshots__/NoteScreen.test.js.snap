// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas NoteScreen Debe mostrarse correctamente 1`] = `
<div
  className="note__main-content"
>
  <div
    className="note__app-bar"
  >
    <span>
      28 agosto 2020
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="note__content"
  >
    <input
      autoComplete="off"
      className="note__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Input"
      type="text"
      value="Hola mundo"
    />
    <textarea
      className="note__textarea"
      name="body"
      onChange={[Function]}
      placeholder="Escribe aquÃ­"
      value="Soy el cuerpo"
    />
  </div>
  <div
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </div>
</div>
`;
